### PR TITLE
Feature/wjiang/large output

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
When the output file is large, the size of the array should be declared as INT64. More explicit error message is printed out when data for MPI send is too big 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
This PR addresses #1942 . The failed message would be more specific now

File: pfio/ForwardDataAndMessage.F90 [hunk 1]
```diff
@@ -56,7 +56,9 @@ subroutine serialize(this, buffer, rc)
       else
          buffer = buff_tmp
       endif
-
+      if ( size(buffer, kind=8) > huge(0)) then
+        _ASSERT(.false., "need to increase oserver's number of front cores (nfront)")
```
Comment (mathomp4) [inline pfio/ForwardDataAndMessage.F90:60]:
```suggestion
        _FAIL("need to increase oserver's number of front cores (nfront)")
```

File: pfio/MessageVector.F90 [hunk 2]
```diff
@@ -42,7 +43,13 @@ subroutine serialize_message_vector(msgVec,buffer)
         msg=>msgVec%at(i)
         tmp =[tmp, parser%encode(msg)]
      enddo
+
+     if(size(tmp, kind=8) > huge(0)) then
+       _ASSERT( .false., " need to increase oserver's nfront")
```
Comment (mathomp4) [inline pfio/MessageVector.F90:48]:
```suggestion
       _FAIL("need to increase oserver's nfront")
```

**Comment by metdyn** [comment]
@weiyuan-jiang,  I find another segmentation fault related to INT_4 error using this branch for writing C2880 with 181 layer on 2 oserver node (2 compute node). I used gcc12. I am not completely sure if this is due to memory. Please check my directory.

```
 MultiServer Start: nfront, nwriter          54          37
Starting pFIO output server on 2 nodes
 Running new case
     SHMEM: NumCores per Node varies from 8 to 46
     SHMEM: NumNodes in use   = 2
     SHMEM: Total PEs         = 54

  Output RSLV:         2880       17280
 Writing:    181 Slices to File:  case1.2004.nc4


AGCM Date: 2004/01/15  Time: 21:00:00   37.5%Memory Committed
[borgg104:43458:0:43458] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace (tid:  43458) ====
 0  /usr/lib64/libucs.so.0(ucs_handle_error+0xe4) [0x2b7c11501da4]
 1  /usr/lib64/libucs.so.0(+0x2210c) [0x2b7c1150210c]
 2  /usr/lib64/libucs.so.0(+0x222c2) [0x2b7c115022c2]
 3  /lib64/libpthread.so.0(+0x11ce0) [0x2b7bffe57ce0]
 4  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.profiler.so(__mapl_abstractcolumn_MOD___copy_INTEGER_4_+0x16) [0x2b7bfbc3f6a0]
 5  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_tt_resize_size+0x13c3) [0x2b7bfae3b7f9]
 6  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_tt_push_back+0xb1) [0x2b7bfae416ff]
 7  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_s_insert+0x2f1) [0x2b7bfae4ed61]
 8  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_m_insert_key_value+0xc07) [0x2b7bfae4b6b5]
 9  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(+0x363e19) [0x2b7bfb02ce19]
10  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_multigroupservermod_MOD_start_back+0x3db) [0x2b7bfb03c452]
11  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_multigroupservermod_MOD_start+0xa5) [0x2b7bfb0456a8]
12  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/lib/libMAPL.base.so(__mapl_servermanager_MOD_initialize+0xc877) [0x2b7bfa6c1938]
13  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/bin/ExtDataDriver.x() [0x42f665]
14  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/bin/ExtDataDriver.x() [0x42f9a5]
15  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/bin/ExtDataDriver.x() [0x42c36f]
16  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/bin/ExtDataDriver.x() [0x42c3f7]
17  /lib64/libc.so.6(__libc_start_main+0xf5) [0x2b7c004eaac5]
18  /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/build_mapl/bin/ExtDataDriver.x() [0x4096f9]
=================================


Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2b7bffe57cdf in ???
#1  0x2b7bfbc3f6a0 in __mapl_abstractcolumn_MOD___copy_INTEGER_4_
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/profiler/reporting/AbstractColumn.F90:89
#2  0x2b7bfae3b7f8 in __pfio_stringattributemapmod_MOD_tt_resize_size
        at /discover/swdev/gmao_SIteam/Baselibs/ESMA-Baselibs-7.7.0/x86_64-pc-linux-gnu/gfortran_12.1.0-openmpi_4.1.3/Linux/GFTL-1.8/include/v1/templates/vector_impl.inc:497
#3  0x2b7bfae416fe in __pfio_stringattributemapmod_MOD_tt_push_back
        at /discover/swdev/gmao_SIteam/Baselibs/ESMA-Baselibs-7.7.0/x86_64-pc-linux-gnu/gfortran_12.1.0-openmpi_4.1.3/Linux/GFTL-1.8/include/v1/templates/vector_impl.inc:402
#4  0x2b7bfae4ed60 in __pfio_stringattributemapmod_MOD_s_insert
        at /discover/swdev/gmao_SIteam/Baselibs/ESMA-Baselibs-7.7.0/x86_64-pc-linux-gnu/gfortran_12.1.0-openmpi_4.1.3/Linux/GFTL-1.8/include/v1/templates/altSet_impl.inc:300
#5  0x2b7bfae4b6b4 in __pfio_stringattributemapmod_MOD_m_insert_key_value
        at /discover/swdev/gmao_SIteam/Baselibs/ESMA-Baselibs-7.7.0/x86_64-pc-linux-gnu/gfortran_12.1.0-openmpi_4.1.3/Linux/GFTL-1.8/include/v1/templates/map_impl.inc:84
#6  0x2b7bfb02ce18 in start_back_writers
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/pfio/MultiGroupServer.F90:669
#7  0x2b7bfb03c451 in __pfio_multigroupservermod_MOD_start_back
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/pfio/MultiGroupServer.F90:447
#8  0x2b7bfb0456a7 in __pfio_multigroupservermod_MOD_start
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/pfio/MultiGroupServer.F90:201
#9  0x2b7bfa6c1937 in __mapl_servermanager_MOD_initialize
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/base/ServerManager.F90:265
#10  0x42f664 in __extdatadrivermod_MOD_initialize_io_clients_servers
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/Tests/ExtDataDriverMod.F90:183
#11  0x42f9a4 in __extdatadrivermod_MOD_run
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/Tests/ExtDataDriverMod.F90:91
#12  0x42c36e in extdata_driver
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/Tests/ExtDataDriver.F90:25
#13  0x42c3f6 in main
        at /discover/nobackup/yyu11/br_mapl_INT64_pfio_gcc/MAPL/Tests/ExtDataDriver.F90:6
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
```

**Comment by weiyuan-jiang** [comment]
This error probably from the bug here https://github.com/Goddard-Fortran-Ecosystem/gFTL/blob/24f2370880cfac4f27781b473303f6fc0b328456/include/v1/templates/vector_impl.inc#L56 . It should be size(this%elements,kind=SIZE_KIND)

Need @tclune to confirm this.

**Comment by mathomp4** [comment]
> This error is probably caused by the bug here https://github.com/Goddard-Fortran-Ecosystem/gFTL/blob/24f2370880cfac4f27781b473303f6fc0b328456/include/v1/templates/vector_impl.inc#L56 . It should be size(this%elements,kind=SIZE_KIND)
> 
> Need @tclune to confirm this.

@weiyuan-jiang Do you know if there is a similar issue in v2 as well? I tried taking a look but gFTL still baffles me at times. So complex!

I think maybe there is as this code: 

https://github.com/Goddard-Fortran-Ecosystem/gFTL/blob/24f2370880cfac4f27781b473303f6fc0b328456/include/v2/vector/procedures.inc#L132-L145

looks similar.

**Comment by tclune** [comment]
Indeed.  Good catch.

@weiyuan - please issue a hotfix PR on the gFTL repo.     In the mean time, you could look at switching to gFTL2 which probably does not have the issue.    (It will take time to propagate the hotfix up to a new release of baselibs.)
@mathomp4  - this unfortunately has some priority.

**Comment by tclune** [comment]
I see that @weiyuan-jiang already proactively submitted a PR for gFTL.    I've fixed that up and issued a new release of gFTL (1.8.3). 

@mathomp4 maybe you could do a "hack" installation of Baselibs to let @metdyn proceed in the mean time rather than waiting for proper updates to GFE?

**Comment by mathomp4** [comment]
> I see that @weiyuan-jiang already proactively submitted a PR for gFTL. I've fixed that up and issued a new release of gFTL (1.8.3).
> 
> @mathomp4 maybe you could do a "hack" installation of Baselibs to let @metdyn proceed in the mean time rather than waiting for proper updates to GFE?

Sure. I want to do one with the yaFyaml changes as well. I'll get to work on it now.

**Comment by tclune** [comment]
It is of course possible that there is a similar problem in V2.  But I had a much more consistent approach to this issue when doing V2, so it is somewhat less likely.   I'll take a quick scan.

**Comment by tclune** [comment]
I take it back.   I did spot several potential issues due to using default integers.    Definitely need to do a PR.    Need to check other containers as well, but probably isolated to Vector.

**Comment by weiyuan-jiang** [comment]
More fixes are in this PR https://github.com/Goddard-Fortran-Ecosystem/gFTL/pull/190

**Comment by weiyuan-jiang** [comment]
@metdyn I suspected that this is still a memory issue.  During the peak time, there are 3 copies of large array (or the large file) in the memory. I tried to reduce it to 2 copies now. Would you please give it another try ?

**Comment by metdyn** [comment]
@weiyuan-jiang (@tclune):

Good job for destroying a copy of the variable in memory!! 
C2880  LZ=181 now writes one 3dvar using 2 computer node and 2 io nodes (4 nodes total).  Write out nc4 file with 36 GB data.   gcc12 with openmpi on discover (thank @mathomp4).  The latest PR branches for gFTL and MPAL are used.  Let me try C5760 to give you more feedback.

 MultiServer Start: nfront, nwriter          54          37
Starting pFIO output server on 2 nodes
 Running new case
     SHMEM: NumCores per Node varies from 8 to 46
     SHMEM: NumNodes in use   = 2
     SHMEM: Total PEs         = 54
     SHMEM: NumNodes in use  = 2
  oserver is not split

       Slices:          181
  Output RSLV:         2880       17280
 Writing:    181 Slices to File:  case1.2004.nc4 
36G Jan 31 12:07 case1.2004.nc4

**Comment by tclune** [comment]
Is the 3D field being split?   Or why are there 2 I/O nodes?   I could just be out of date - maybe 1 node is just to manage the round-robin?

**Comment by tclune** [comment]
I meant to say "Excellent News!" at the beginning of my previous message.

**Comment by weiyuan-jiang** [comment]
> 

When it comes to a big file, it is safer to have more front cores than usual  because the front cores would combine the data it receives from applications and forward it to the writer.   The combined data may exceed the maximum length for MPI_send if the number of front cores is not big enough.

**Comment by metdyn** [comment]
@tclune:  The 181 layers were not split in writing; they were written as a whole!   That is excellent news!  (and I am hoping to see the C5760 writing.)

In align with what @weiyuan-jiang mentioned, the 2 io-node were used in a way that the number of front cores equal the number of compute cores (54 cores in this case). 
The output shows `AGCM Date: 2004/01/15  Time: 21:00:00   37.5%Memory Committed`; I feel it is a good bound for memory usage.  I think using 1 io-node may also work. I will give it a try!

**Comment by metdyn** [comment]
@weiyuan-jiang:

As @tclune suggested to use 1 io node to test C2880.  I used 2 compute nodes with 1 io nodes.  There appears another `mapl_abstractcolumn_MOD___copy_INTEGER_4`  segmentation fault.  It is different from memory allocation error.   Please check my directory. Thanks!
```
 MultiServer Start: nfront, nwriter          40           5
Starting pFIO output server on 1 nodes
 Running new case
     SHMEM: NumCores per Node varies from 8 to 46
     SHMEM: NumNodes in use   = 2
     SHMEM: Total PEs         = 54
     SHMEM: NumNodes in use  = 2


  Output RSLV:         2880       17280


 Writing:    181 Slices to File:  case1.2004.nc4

 AGCM Date: 2004/01/15  Time: 21:00:00   37.5%Memory Committed
[borgi100:36399:0:36399] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace (tid:  36399) ====
 0  /usr/lib64/libucs.so.0(ucs_handle_error+0xe4) [0x2b78a0efcda4]
 1  /usr/lib64/libucs.so.0(+0x2210c) [0x2b78a0efd10c]
 2  /usr/lib64/libucs.so.0(+0x222c2) [0x2b78a0efd2c2]
 3  /lib64/libpthread.so.0(+0x11ce0) [0x2b788f8d4ce0]
 4  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.profiler.so(__mapl_abstractcolumn_MOD___copy_INTEGER_4_+0x16) [0x2b788b8df679]
 5  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_tt_resize_size+0x13c3) [0x2b788aae152d]
 6  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_tt_push_back+0xae) [0x2b788aae7430]
 7  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_s_insert+0x2f1) [0x2b788aaf4a94]
 8  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_stringattributemapmod_MOD_m_insert_key_value+0xc07) [0x2b788aaf13e2]
 9  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(+0x364fb8) [0x2b788acd2fb8]
10  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_multigroupservermod_MOD_start_back+0x3db) [0x2b788ace2616]
11  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.pfio.so(__pfio_multigroupservermod_MOD_start+0xa5) [0x2b788aceb86c]
12  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/lib/libMAPL.base.so(__mapl_servermanager_MOD_initialize+0xc877) [0x2b788a36db60]
13  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/bin/ExtDataDriver.x() [0x42f657]
14  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/bin/ExtDataDriver.x() [0x42f997]
15  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/bin/ExtDataDriver.x() [0x42c361]
16  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/bin/ExtDataDriver.x() [0x42c3e9]
17  /lib64/libc.so.6(__libc_start_main+0xf5) [0x2b788ff67ac5]
18  /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/build_mapl/bin/ExtDataDriver.x() [0x4096d9]
=================================



Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2b788f8d4cdf in ???
#1  0x2b788b8df679 in __mapl_abstractcolumn_MOD___copy_INTEGER_4_
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/profiler/reporting/AbstractColumn.F90:89
#2  0x2b788aae152c in __pfio_stringattributemapmod_MOD_tt_resize_size
        at /discover/nobackup/yyu11/tools/bs2.gcc12/x86_64-pc-linux-gnu/gfortran/Linux/GFTL-1.8/include/v1/templates/vector_impl.inc:497
#3  0x2b788aae742f in __pfio_stringattributemapmod_MOD_tt_push_back
        at /discover/nobackup/yyu11/tools/bs2.gcc12/x86_64-pc-linux-gnu/gfortran/Linux/GFTL-1.8/include/v1/templates/vector_impl.inc:402
#4  0x2b788aaf4a93 in __pfio_stringattributemapmod_MOD_s_insert
        at /discover/nobackup/yyu11/tools/bs2.gcc12/x86_64-pc-linux-gnu/gfortran/Linux/GFTL-1.8/include/v1/templates/altSet_impl.inc:300
#5  0x2b788aaf13e1 in __pfio_stringattributemapmod_MOD_m_insert_key_value
        at /discover/nobackup/yyu11/tools/bs2.gcc12/x86_64-pc-linux-gnu/gfortran/Linux/GFTL-1.8/include/v1/templates/map_impl.inc:84
#6  0x2b788acd2fb7 in start_back_writers
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/pfio/MultiGroupServer.F90:672
#7  0x2b788ace2615 in __pfio_multigroupservermod_MOD_start_back
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/pfio/MultiGroupServer.F90:447
#8  0x2b788aceb86b in __pfio_multigroupservermod_MOD_start
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/pfio/MultiGroupServer.F90:201
#9  0x2b788a36db5f in __mapl_servermanager_MOD_initialize
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/base/ServerManager.F90:265
#10  0x42f656 in __extdatadrivermod_MOD_initialize_io_clients_servers
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/Tests/ExtDataDriverMod.F90:183
#11  0x42f996 in __extdatadrivermod_MOD_run
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/Tests/ExtDataDriverMod.F90:91
#12  0x42c360 in extdata_driver
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/Tests/ExtDataDriver.F90:25
#13  0x42c3e8 in main
        at /discover/nobackup/yyu11/br_mapl_wj_Jan25_gcc/MAPL/Tests/ExtDataDriver.F90:6
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
```

**Comment by weiyuan-jiang** [comment]
The symptom looks same as before. So it is still a memory issue. Let me try to reduce from 2 copies to one copy and see what happens

**Comment by tclune** [comment]
@metdyn To be clear, I was just asking _why_ 2 nodes were used.  @weiyuan-jiang answered that.   I do not think it is important to try with 1 node.

Also, please keep in mind that there is no pressing need for the _History_ component to output 3D fields.  It will be useful to know, but only the 2D outputs are important for now.    3D full-resolution fields will come from the checkpoint/restart layer of GEOS, but that requires use of the full model as opposed to the mechanism you are testing with for now.  (And of course being able to do 3D full-resolution in History would be a wonderful capability.   I'm just fairly confident we won't achieve that in the short run.)

Garrison's work will hopefully supplement our understanding of the scalability limits here.

**Comment by metdyn** [comment]
Hi @tclune:  fully understand your vision and the scenario!  Well, @weiyuan-jiang may have  another solution to even make this demanding task to work on history output. Either way, I am glad this code can reach some upper boundary for memory efficiency. That makes my efforts in testing the work meaningful! Thank you for letting me know about the 2d/3d vs history/restart scenario, @bena-nasa also mentioned to me about this.  We would intend to supplement true runs via probing first the "atomic limit" ...   Thank you!

> @metdyn To be clear, I was just asking _why_ 2 nodes were used. @weiyuan-jiang answered that. I do not think it is important to try with 1 node.
> 
> Also, please keep in mind that there is no pressing need for the _History_ component to output 3D fields. It will be useful to know, but only the 2D outputs are important for now. 3D full-resolution fields will come from the checkpoint/restart layer of GEOS, but that requires use of the full model as opposed to the mechanism you are testing with for now. (And of course being able to do 3D full-resolution in History would be a wonderful capability. I'm just fairly confident we won't achieve that in the short run.)
> 
> Garrison's work will hopefully supplement our understanding of the scalability limits here.

**Review by mathomp4 — APPROVED** [review]

**Pull request merged**